pecos/core/utils/clustering.hpp

```diff
@@ -155,14 +203,46 @@ struct Tree {
     }
 
     template<typename MAT>
-    void partition_kmeans(size_t nid, size_t depth, const MAT& feat_mat, rng_t& rng, size_t max_iter=10, int threads=1, int thread_id=0) {
-        Node& root = root_of(nid);
-        Node& left = left_of(nid);
-        Node& right = right_of(nid);
+    void partition_kmeans(size_t nid, size_t depth, const MAT& feat_mat, rng_t& rng, size_t max_iter=10, int threads=1, int thread_id=0, float cur_sample_rate=1.0) {
+        // copy nodes rather than reference for sampling
+        Node root = root_of(nid);
+        Node left = left_of(nid);
+        Node right = right_of(nid);
+
+        // modify nodes' start and end based on cur_sample_rate 
+        size_t n_sp_elements = 0;
+        if(cur_sample_rate < 1.0) {
```

Comment:
the following 4 lines of code will also be used in `partition_skmeans`. Consider refactor it into functions, so it can be called twice, reducing code redundancy.